shared/p2p/hobbits/server.go

```diff
@@ -0,0 +1,190 @@
+package hobbits
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"reflect"
+	"strconv"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared"
+	"github.com/prysmaticlabs/prysm/shared/event"
+	"github.com/prysmaticlabs/prysm/shared/p2p"
+	"github.com/renaynay/go-hobbits/encoding"
+	log "github.com/sirupsen/logrus"
+	ttl "github.com/ReneKroon/ttlcache"
+	"gopkg.in/mgo.v2/bson"
+)
+
+func NewHobbitsNode(host string, port int, peers []string, db *db.BeaconDB) HobbitsNode {
+	cache := ttl.NewCache()
+	cache.Set(string(make([]byte, 32)), true)
+
+	return HobbitsNode{
+		NodeId:       strconv.Itoa(rand.Int()),
+		Host:         host,
+		Port:         port,
+		StaticPeers:  peers,
+		PeerConns:    make(map[peer.ID]net.Conn),
+		feeds:        make(map[reflect.Type]p2p.Feed),
+		DB:           db,
+		MessageStore: cache,
+	}
+}
+
+func (h *HobbitsNode) OpenConns() error {
+	for _, p := range h.StaticPeers {
+		go func(p string) {
+			var conn net.Conn
+			var err error
+
+			for i := 0; i <= 10; i++ {
+				conn, err = net.DialTimeout("tcp", p, 3*time.Second)
+				if err == nil {
+					break
+				}
+
+				fmt.Println(err)
+
+				time.Sleep(5 * time.Second)
+			}
+
+			h.Lock()
+
+			h.PeerConns[peer.ID(p)] = conn
+
+			h.Unlock()
+		}(p)
+	}
+
+	return nil
+}
+
+func (h *HobbitsNode) Listen() error {
+	log.Trace("hobbits node is listening")
+
+	return h.Server.Listen(func(conn net.Conn, message encoding.Message) {
+		id := peer.ID(conn.RemoteAddr().String())
+		_, ok := h.PeerConns[id]
+		if !ok {
+			h.PeerConns[id] = conn
+		}
+
+		err := h.processHobbitsMessage(id, HobbitsMessage(message))
+		if err != nil {
+			log.Trace(fmt.Sprintf("closing conn and deleting peer: %s", h.PeerConns[id].RemoteAddr().String()))
+			log.Error(err)
+
+			_ = conn.Close()
+			delete(h.PeerConns, id)
+
```

Comment:
again , this is vulnerable to concurrent writes, and remove the empty line below